language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/TemplateClass.scala

```diff
@@ -62,116 +62,183 @@ private[inner] object TemplateClass extends StrictLogging {
-      .addMethod(generateFromIdAndRecord(contractClassName, templateClassName, contractIdClassName))
+      .addMethod(
+        generateFromIdAndRecord(
+          contractClassName,
+          templateClassName,
+          contractIdClassName,
+          contractKeyClassName))
       .addMethod(
         generateFromIdAndRecordDeprecated(
           contractClassName,
           templateClassName,
-          contractIdClassName))
+          contractIdClassName,
+          contractKeyClassName.isDefined))
       .addMethod(
-        generateFromCreatedEvent(contractClassName, templateClassName, contractIdClassName))
+        generateFromCreatedEvent(
+          contractClassName,
+          templateClassName,
+          contractIdClassName,
+          contractKeyClassName))
       .addMethods(ObjectMethods(contractClassName, fields, templateClassName).asJava)
       .build()
   }
 
   private[inner] def generateFromIdAndRecord(
       className: ClassName,
       templateClassName: ClassName,
-      idClassName: ClassName): MethodSpec =
-    MethodSpec
-      .methodBuilder("fromIdAndRecord")
-      .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
-      .returns(className)
-      .addParameter(classOf[String], "contractId")
-      .addParameter(classOf[javaapi.data.Record], "record$")
-      .addParameter(optionalString, agreementFieldName)
-      .addParameter(optionalValue, contractKeyFieldName)
-      .addStatement("$T $L = new $T(contractId)", idClassName, idFieldName, idClassName)
-      .addStatement(
-        "$T $L = $T.fromValue(record$$)",
-        templateClassName,
-        dataFieldName,
-        templateClassName)
-      .addStatement(
+      idClassName: ClassName,
+      maybeContractKeyClassName: Option[TypeName]): MethodSpec = {
+
+    val params = Iterable(
+      ParameterSpec.builder(classOf[String], "contractId").build(),
+      ParameterSpec.builder(classOf[javaapi.data.Record], "record$").build(),
+      ParameterSpec.builder(optionalString, agreementFieldName).build()
+    ) ++ maybeContractKeyClassName
+      .map(name => ParameterSpec.builder(optional(name), contractKeyFieldName).build)
+      .toList
+
+    val spec =
+      MethodSpec
+        .methodBuilder("fromIdAndRecord")
+        .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+        .returns(className)
+        .addParameters(params.asJava)
+        .addStatement("$T $L = new $T(contractId)", idClassName, idFieldName, idClassName)
+        .addStatement(
+          "$T $L = $T.fromValue(record$$)",
+          templateClassName,
+          dataFieldName,
+          templateClassName)
+
+    if (maybeContractKeyClassName.isDefined) {
+      spec.addStatement(
         "return new $T($L, $L, $L, $L)",
         className,
         idFieldName,
         dataFieldName,
         agreementFieldName,
-        contractKeyFieldName
-      )
-      .build()
+        contractKeyFieldName)
+    } else {
+      spec.addStatement(
+        "return new $T($L, $L, $L)",
```

Comment:
You could build up the return statement with something like the following to avoid having to duplicate the statement itself:
```
val names = Vector(idFieldName, dataFieldName, agreementFieldName) ++ maybeContractKeyClassName.map(_ => contractKeyFieldName)
val params = names.map(n => CodeBlock.of("$L", n)).asJava
val allParams = CodeBlock.join(params, ", ")
spec.addStatement("return new $T($L)", className, params)
```

Reply:
Thanks, both my Scala and my `javapoet` became rusty after the last few months. :slightly_smiling_face:

Reply:
Addressed by https://github.com/digital-asset/daml/pull/1586/commits/42493732637d847f3a881a533ac2d387bd5d14f9